src/types.ts

```diff
@@ -0,0 +1,137 @@
+/**
+ * These represent the types in our abstract type system.
+ * We will normalize the raw XTP schema into these recursively defined types.
+ */
+
+export type XtpNormalizedKind =
+  'object' | 'enum' | 'map' | 'array' | 'string' |
+  'int32' | 'int64' | 'float' | 'double' |
+  'boolean' | 'date-time' | 'byte' | 'buffer'
+
+// applies type opts to a type on construction
+function cons(t: XtpNormalizedType, opts?: XtpTypeOpts): XtpNormalizedType {
+  // default them to false
+  t.nullable = (opts?.nullable === undefined) ? false : opts.nullable
+  t.required = (opts?.required === undefined) ? false : opts.required
+  return t
+}
+
+export interface XtpTypeOpts {
+  nullable?: boolean;
+  required?: boolean;
+}
+
+export interface XtpNormalizedType extends XtpTypeOpts {
+  kind: XtpNormalizedKind;
+}
+
+export class StringType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'string';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class Int32Type implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'int32';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class Int64Type implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'int64';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class FloatType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'float';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class DoubleType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'double';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class BooleanType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'boolean';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class ByteType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'byte';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class BufferType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'buffer';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class DateTimeType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'date-time';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class ObjectType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'object';
+  name: string;
+  properties: Array<XtpNormalizedType>;
+
+  constructor(name: string, properties: Array<XtpNormalizedType>, opts?: XtpTypeOpts) {
```

Comment:
Some of these more complex types can now be recursively defined. this gives us easier interface to generate code and signatures from.